rosidl_runtime_c/src/type_description_utils.c

```diff
@@ -0,0 +1,1514 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// NOTE(methylDragon): Maybe this belongs inside the type_description_interfaces
+//                     package instead?
+//                     But as a C header (.h)???
+
+// NOTE(methylDragon): I made it so that every instance of a non-message struct (e.g. hash map)
+//                     borrows, whereas the message structs copy.
+//                     So lifetime should be managed by the message structs.
+
+#include "rosidl_runtime_c/type_description_utils.h"
+
+#include <inttypes.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <rcutils/logging_macros.h>
+#include <rcutils/repl_str.h>
+#include <rcutils/qsort.h>
+#include <rcutils/types/rcutils_ret.h>
+#include <rcutils/types/hash_map.h>
+#include <rosidl_runtime_c/string_functions.h>
+
+#include "rosidl_runtime_c/type_description/field__functions.h"
+#include "rosidl_runtime_c/type_description/field__struct.h"
+#include "rosidl_runtime_c/type_description/individual_type_description__functions.h"
+#include "rosidl_runtime_c/type_description/individual_type_description__struct.h"
+#include "rosidl_runtime_c/type_description/type_description__functions.h"
+#include "rosidl_runtime_c/type_description/type_description__struct.h"
+
+
+// Modified from http://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2
+size_t next_power_of_two(size_t v)
+{
+  v--;
+  size_t shf = 1;
+  v |= v >> 1;
+  for (size_t i = 0; i < sizeof(size_t); i++) {
+    shf = shf * 2;
+    v |= v >> shf;
+  }
+  v++;   // next power of 2
+  return v > 1 ? v : 1;
+}
+
+// =================================================================================================
+// GET BY NAME
+// =================================================================================================
+
+rcutils_ret_t
+rosidl_runtime_c_type_description_utils_find_field(
+  const rosidl_runtime_c__type_description__Field__Sequence * fields,
+  const char * name,
+  rosidl_runtime_c__type_description__Field ** field)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(fields, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(name, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(field, RCUTILS_RET_INVALID_ARGUMENT);
+  if (*field != NULL) {
+    RCUTILS_LOG_ERROR("`field` output argument is not pointing to null");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+
+  for (size_t i = 0; i < fields->size; i++) {
+    if (strcmp(fields->data[i].name.data, name) == 0) {
```

Comment:
Do mesage names have a maximum length? I didn't see a limit...

Reply:
I don't think so, actually now that I think about it. It was more of a "hey did you consider this" rather than a rhetorical question that I already had an answer to. So I'm fine leaving this as-is.